### PR TITLE
New Presence "AniCloud"

### LINKED ISSUE #3530: AniCloud | AniCloud.io
**Prerequisites and essential questions** 
- [x] Is it a popular site?
- [x] Is the website older than 2 months? 
- [ ] Is the site locked to a specific country/region?
- [ ] Is the site a paid service? (e.g. Netflix, Hulu)
- [ ] Does the website feature NSFW content? (e.g. porn, etc...)
- [ ] Are you a donator/patron?
- [x] Do you acknowledge that coding presences is completely voluntary and may take time for your service to be added regardless of priority?

**What's your Discord username?** **I мΛχ I#6832**

**What's the name of the service?** **www.anicloud.io**

**What should the Presence display?** **what the person is watching (Anime+Season+Episode) and the time that is left**

**If possible, please provide a logo for the service (512x512 minimum)** 
AniCloud io

### PR DESCRIPTION
**Pull request by Emijor**
Resolves #3530 

image

image

image

image

image

### CODE REVIEW

File: websites/A/Anicloud/presence.ts [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+* @param {Number} videoDuration Video duration seconds
+*/
+function getTimestamps(
+videoTime: number,
+videoDuration: number
+): Array<number> {
+const startTime = Math.floor(Date.now() / 1000),
+  endTime = Math.floor(startTime - videoTime + videoDuration);
+return [startTime, endTime];
+}
```
Comment (SlowLife1661) [inline websites/A/Anicloud/presence.ts:23]: Use built-in method, Presence.getTimestamps()

File: websites/A/Anicloud/iframe.ts [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+      const played = video.currentTime != 0;
+      iframe.send({
+        current: video.currentTime,
+        duration: video.duration,
+        paused: video.paused,
+        played
+      });
+    }
+  }
+});
```
Comment (SlowLife1661) [inline websites/A/Anicloud/iframe.ts:21]: Delete

File: websites/A/Anicloud/dist/metadata.json [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+    "version": "1.0.1",
+    "logo": "https://i.imgur.com/Ex3XUUc.png",
+    "thumbnail": "https://i.imgur.com/fKBCijM.png",
+    "color": "#637cf9",
+    "tags": [
+      "anime",
+      "videos"
+    ],
+    "category": "anime",
+    "iframe": true
```
Comment (SlowLife1661) [inline websites/A/Anicloud/dist/metadata.json:22]: Delete

File: websites/A/Anicloud/dist/metadata.json [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+      "id": "466885774621540372"
+    },
+    "service": "AniCloud",
+    "description": {
+      "en": "Watch now over 1,000+ animes for free on AniCloud. Anime Video On Demand: ✓ 100% Free ✓ Instantly Online ✓ Exclusive Animes",
+      "de": "Jetzt über 1.000+ Animes als Stream kostenlos bei AniCloud ansehen. Anime Video-On-Demand: ✓ 100% Kostenlos ✓ Sofort online ✓ Exclusive Animes"
+    },
+    "regExp": "([a-z0-9-]+[.])*anicloud([.][a-z]+)+[/]",
+    "url": "anicloud.io",
+    "version": "1.0.1",
```
Comment (SlowLife1661) [inline websites/A/Anicloud/dist/metadata.json:13]: 1.0.0

File: websites/A/Anicloud/presence.ts [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+  duration: number;
+  paused: boolean;
+  played: boolean;
+}) => {
+  current = data.current;
+  duration = data.duration;
+  paused = data.paused;
+  played = data.played;
+}
+);
```
Comment (SlowLife1661) [inline websites/A/Anicloud/presence.ts:44]: Delete, you aren't using iframe

File: websites/A/Anicloud/presence.ts [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+const startTime = Math.floor(Date.now() / 1000),
+  endTime = Math.floor(startTime - videoTime + videoDuration);
+return [startTime, endTime];
+}
+
+let video,
+current: number,
+duration: number,
+paused: boolean,
+played: boolean;
```
Comment (SlowLife1661) [inline websites/A/Anicloud/presence.ts:29]: Remove video you aren't using it, you declared it 2 times lol

File: websites/A/Anicloud/presence.ts [hunk 7]
```diff
@@ -0,0 +1,235 @@
...
+
+if (document.location.pathname.startsWith("/anime/")) {
+  const title = document.querySelector("h1").textContent,
+  ep = document.querySelector("#wrapper > div.seriesContentBox > div.container.marginBottom > ul > li.currentActiveLink > a > span").textContent.match(/Episode\W\d/),
+  actvie = document.querySelector("h2").textContent;
+    presenceData.details = title;
+    presenceData.state = ep + " | " + actvie;
+    presenceData.buttons = [
+      {
+        label: "Current Anime",
```
Comment (SlowLife1661) [inline websites/A/Anicloud/presence.ts:59]: Watch Episode?

File: websites/A/Anicloud/presence.ts [hunk 8]
```diff
@@ -0,0 +1,235 @@
...
+    duration = video.duration;
+    current = video.currentTime;
+    paused = video.paused;
+  }
+  if (played) {
+    if (!paused) {
+      const timestamps = getTimestamps(current, duration);
+      presenceData.startTimestamp = timestamps[0];
+      presenceData.endTimestamp = timestamps[1];
+    }
```
Comment (SlowLife1661) [inline websites/A/Anicloud/presence.ts:76]:
```suggestion
  if (video) {
    if (!video.paused) {
      const timestamps = presence.getTimestamps(video.currentTime, video.duration);
      presenceData.endTimestamp = timestamps[1];
    }
```

**Review by StrikerFRFX — CHANGES_REQUESTED** [review]
Hi @Emijor, 
I've requested changes below in my review which you should take a look at 😄 
 -  In addition to changes that have been suggested by another user, fix Codacy.

Thanks,
**Striker**, PreMiD Reviewer Team.
Contact: me@strikerfrfx.dev, Striker#1337

**Comment by Emijor** [comment]
wait... if I delete iframe then my presence no longer works. I dont get it

File: websites/A/Anicloud/presence.ts (hunk 8 above)
Comment (SlowLife1661) [inline websites/A/Anicloud/presence.ts:76]:
```suggestion
    video = document.querySelector(`video`);
  if (video != null) {
    played = video.currentTime != 0;
    duration = video.duration;
    current = video.currentTime;
    paused = video.paused;
  }
  if (played) {
    if (!paused) {
      const timestamps = presence.getTimestamps(current, duration);
      presenceData.endTimestamp = timestamps[1];
    }
```

**Review by Bas950 — APPROVED** [review]
Domain is just old enough lmao xD

File: websites/A/Anicloud/iframe.ts [hunk 9]
```diff
@@ -0,0 +1,20 @@
+const iframe = new iFrame();
+iframe.on("UpdateData", async () => {
+  const { hostname } = window.location;
+  if (
+    hostname === `voe.sx` ||
+    hostname === `streamtape.com` ||
+    hostname == `playtube.ws`
```
Comment (ririxidev) [inline websites/A/Anicloud/iframe.ts:7]:
```suggestion
    hostname === "voe.sx" ||
    hostname === "streamtape.com" ||
    hostname === "playtube.ws"
```

File: websites/A/Anicloud/presence.ts [hunk 10]
```diff
@@ -0,0 +1,219 @@
+const presence = new Presence({
+  clientId: `830517484472762408`
+}),
+
```
Comment (ririxidev) [inline websites/A/Anicloud/presence.ts:4]: dont need blank space here?

**Review by ririxidev — CHANGES_REQUESTED** [review]

**Comment by Emijor** [comment]
like this?

File: websites/A/Anicloud/presence.ts [hunk 11]
```diff
@@ -0,0 +1,220 @@
+const presence = new Presence({
+  clientId: `830517484472762408`
+
+}),
+
+  strings = presence.getStrings({
+  play: "presence.playback.playing",
+  pause: "presence.playback.paused"
+});
```
Comment (ririxidev) [inline websites/A/Anicloud/presence.ts:9]:
```suggestion
const presence = new Presence({
  clientId: `830517484472762408`
}),
  strings = presence.getStrings({
  play: "presence.playback.playing",
  pause: "presence.playback.paused"
});
```

**Review by ririxidev — CHANGES_REQUESTED** [review]

File: websites/A/Anicloud/presence.ts (hunk 11 above)
Comment (Emijor) [reply websites/A/Anicloud/presence.ts:9]: oh my bad

**Review by ririxidev — APPROVED** [review]

**Review by Bas950 — APPROVED** [review]

**Pull request merged**